FluentStorage.SFTP/SshNetSftpBlobStorage.cs

```diff
@@ -266,18 +266,43 @@ public SshNetSftpBlobStorage(string host, string username, params PrivateKeyFile
 			SftpClient client = GetClient();
 
 			var folder = StoragePath.Combine(RootDirectory, StoragePath.Normalize(options.FolderPath));
-			IEnumerable<SftpFile> directoryContents = await client.ListDirectoryAsync(folder);
-
-			List<Blob> blobCollection = directoryContents
-			   .Where(dc => (options.FilePrefix == null || dc.Name.StartsWith(options.FilePrefix))
-							&& (dc.IsDirectory || dc.IsRegularFile || dc.OwnerCanRead)
-							&& !cancellationToken.IsCancellationRequested
-							&& dc.Name != "."
-							&& dc.Name != "..")
-			   .Take(options.MaxResults.Value)
-			   .Select(ConvertSftpFileToBlob)
-			   .Where(options.BrowseFilter)
-			   .ToList();
+
+			List<Blob> blobCollection = new List<Blob>();
+			object lockObject = new object();
+
+			async Task ListDirectoryAsync(string folderToList) {
+				IEnumerable<SftpFile> directoryContents = await client.ListDirectoryAsync(folderToList);
+				var tempBlobCollection = directoryContents
+					.Where(dc => (options.FilePrefix == null || dc.Name.StartsWith(options.FilePrefix))
+								 && (dc.IsDirectory || dc.IsRegularFile || dc.OwnerCanRead)
+								 && !cancellationToken.IsCancellationRequested
+								 && dc.Name != "."
+								 && dc.Name != "..")
+					.Take(options.MaxResults.Value)
+					.Select(ConvertSftpFileToBlob)
+					.Where(options.BrowseFilter).ToList();
+
+				lock (lockObject) {
+					blobCollection.AddRange(tempBlobCollection);
+				}
+
+				if (options.Recurse == true) {
+					IEnumerable<string> subFoldersToList = tempBlobCollection
+						.Where(x => x.IsFolder == true)
+						.Select(x => x.FullPath);
+#if NET6_0
+					await Parallel.ForEachAsync(subFoldersToList, async (subFolder, token) => {
+						await ListDirectoryAsync(subFolder);
+					});
+#else
+					foreach (string subFolder in subFoldersToList) { 
+						await ListDirectoryAsync(subFolder);
+					}
+#endif
+				}
+			}
```

Comment:
extract this to a new method please

Reply:
can you do this please?

Reply:
Just finsihed for the day. I'll make this my first task tomorrow morning.

Reply:
I moved the method out, bumped SSH.NET to 2023.0.1, and removed Renci.SshNet.Async. With that changed I had to change SftpClient to use ListDirectoryAsync which is an IAsyncEnumerable. I am unsure if I am using it correctly, or if it can be made more efficient. 

Compared to those other test results,

ParallelForEach = 3.0691516 sec
foreach = 17.7257746 sec
IAsyncEnumerable = 1.9855958 sec